File: plugins/skill-intake/scripts/select_diagrams_per_section.js
#!/usr/bin/env node
'use strict';

const fs = require('node:fs');
const path = require('node:path');

const MIN = 1;
const MAX = 3;

function enforce(perSection) {
  const out = {};
  const violations = [];
  for (const sec of Object.keys(perSection)) {
    const arr = Array.isArray(perSection[sec]) ? perSection[sec] : [perSection[sec]];
    if (arr.length < MIN) violations.push({ section: sec, count: arr.length, rule: `min ${MIN}` });
    if (arr.length > MAX) violations.push({ section: sec, count: arr.length, rule: `max ${MAX}` });
    out[sec] = arr.slice(0, MAX);
  }
  return { ok: violations.length === 0, result: out, violations };
}

function main(argv) {
  const file = argv[2];
  if (!file) { process.stderr.write('usage: select_diagrams_per_section.js <map.json>\n'); return 2; }
  let data;
  try { data = JSON.parse(fs.readFileSync(path.resolve(file), 'utf8')); }
  catch (e) { process.stderr.write(`input error: ${e.message}\n`); return 2; }
  const r = enforce(data);
  process.stdout.write(JSON.stringify(r, null, 2) + '\n');
  return r.ok ? 0 : 1;
}

module.exports = { enforce, MIN, MAX };

if (require.main === module) {
  process.exit(main(process.argv));
}

File: plugins/skill-intake/scripts/select_diagrams_per_section.py
#!/usr/bin/env python3
"""Enforce min/max diagram count per section (1..3) and report violations."""
from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any

MIN = 1
MAX = 3


def enforce(per_section: dict[str, Any]) -> dict[str, Any]:
    out: dict[str, list[Any]] = {}
    violations: list[dict[str, Any]] = []
    for sec, val in per_section.items():
        arr = val if isinstance(val, list) else [val]
        if len(arr) < MIN:
            violations.append({'section': sec, 'count': len(arr), 'rule': f'min {MIN}'})
        if len(arr) > MAX:
            violations.append({'section': sec, 'count': len(arr), 'rule': f'max {MAX}'})
        out[sec] = arr[:MAX]
    return {'ok': len(violations) == 0, 'result': out, 'violations': violations}


def main(argv: list[str]) -> int:
    if len(argv) < 1:
        sys.stderr.write('usage: select_diagrams_per_section.py <map.json>\n')
        return 2
    try:
        data = json.loads(Path(argv[0]).resolve().read_text(encoding='utf-8'))
    except Exception as e:
        sys.stderr.write(f'input error: {e}\n')
        return 2
    r = enforce(data)
    sys.stdout.write(json.dumps(r, ensure_ascii=False, indent=2) + '\n')
    return 0 if r['ok'] else 1


if __name__ == '__main__':
    sys.exit(main(sys.argv[1:]))
